### PR TITLE
PCHR-2054: Optimize Manager Calendar

### PR DESCRIPTION
**Pull request by deb1990**
As part of this ticket Manager Leave Calendar code has been optimized to improve performance.

**Improvements**

1. Calendar data reorganised in following order for each contact
```javascript
contact.calendarData = [
  0: [ //dates for January],
  1: [ //dates for February],
  ...
]
```
This is done to eliminate re calculation of month data.

2. LeaveRequests's indexed by Contact and Date, to reduce looping.
3. In `loadCalendar` instead of updating `vm.managedContacts` on completion of each contact's promise. Calendar data saved temporarily in `tempContactData` and once all the promises are resolved, data is transferred to `vm.managedContacts`. This is done so that browser paints the UI only once.
4. Calendar template moved to `manager-leave-calendar-month.html` for better management.
5. `ng-repeat` has been replaced with `ng-hide`
https://github.com/civicrm/civihr/pull/1653/files#diff-e30fb670ba8fb9dbd9ddc7670af3b9acR87
This done to paint calendar elements at once and avoid re paint on each month selection.
6. Only the current month is rendered at load, rest of the months are loaded with an interval of 500ms in the background. This has been done to speed up initial load and also to eliminate rendering when Month Selection is changed.

**Screengrab**

**Performance(Tested with 15 contacts)**
Initial Load - 6 seconds
Clear months - .5 second, but months are loaded sequencially, from next time its instant
Select a single month - .3seconds(Negligible)

### CODE REVIEW

**Comment by AkA84** [comment]
@deb1990 
> PCHR-2054: Optimize calendar code
PCHR-2054: Optimize calendar functions
PCHR-2054: Optimize calendar functions 2

These are the kind of commits that can be squashed together, they just add noise to the commit history

**Comment by AkA84** [comment]
A note about the UI: the loading spinner is shown not just in place of the actual month's calendar, but also in place of the month header (the month name that is). That looks weird, as a user you just see a bunch of spinners with no context to what they are for

The spinner should be shown _below_ the month header, so you can see something like:

>January
> {{spinner}}
>
>February
> {{spinner}}
>
> etc.

This way you can tell what the spinner is for

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 1]
```diff
@@ -39,12 +41,15 @@ define([
       region: null,
       contacts_with_leaves: false
     };
+    vm.months = ['January', 'February', 'March', 'April', 'May', 'June',
+      'July', 'August', 'September', 'October', 'November', 'December'];
     vm.loading = {
       calendar: false,
+      months: vm.months.map(function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: The de-coupling of the loading flag and the actual thing that the flag is referring to is not optimal, it makes it harder to follow and the structure is not as elegant and simple as it could be. Ideally you would have something like this
```js
[{
  label: 'January',
  leaveRequests: [ /* .. */ ],
  loading: true
},
{
  label: 'February',
  leaveRequests: [ /* .. */ ],
  loading: true
}]
```
where on the object itself you are storing the `loading` state, without relying on some other variable loosely related to it

The `vm.loading` property is used to mark the loading state of sections of the page, rather than individual objects like a month is

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 2]
```diff
@@ -305,10 +284,12 @@ define([
      */
     function loadCalendar() {
       var promises = [];
+      tempContactData = angular.copy(vm.managedContacts);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: This is an equivalent, more functional-like version
```js
/**
 * Loads the calendar data
 *
 * @return {Promise}
 */
function loadCalendar() {
  tempContactData = angular.copy(vm.managedContacts);

  return $q.all(vm.managedContacts.map(function (contact, index) {
    return Calendar.get(contact.id, vm.selectedPeriod.id).then(function (calendar) {
      contact.calendarData = setCalendarProps(contact.id, calendar);
      tempContactData[index].calendarData = setCalendarProps(contact.id, calendar);
      showMonthLoader();
    });
  }));
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 3]
```diff
@@ -13,19 +13,21 @@ define([
...
     $log.debug('Component: manager-leave-calendar');
 
     var dayTypes = [],
       publicHolidays = [],
       leaveRequests = [],
+      leaveRequestsIndexed = [],
       leaveRequestStatuses = [],
-      vm = Object.create(this);
+      vm = Object.create(this),
+      tempContactData = [];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Can you explain what this is for?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 4]
```diff
@@ -305,10 +284,12 @@ define([
      */
     function loadCalendar() {
       var promises = [];
+      tempContactData = angular.copy(vm.managedContacts);
       _.each(vm.managedContacts, function (contact, index) {
         promises.push(Calendar.get(contact.id, vm.selectedPeriod.id)
           .then(function (calendar) {
-            vm.managedContacts[index].calendarData = setCalendarProps(vm.managedContacts[index].id, calendar);
+            tempContactData[index].calendarData = setCalendarProps(vm.managedContacts[index].id, calendar);
+            showMonthLoader();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: You are calling this for _each contact_. Is this intentional?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 5]
```diff
@@ -354,10 +335,22 @@ define([
           to: vm.selectedPeriod.end_date
         }
       })
-        .then(function (leaveRequestsData) {
-          leaveRequests = leaveRequestsData.list;
-          return loadCalendar();
+      .then(function (leaveRequestsData) {
+        leaveRequests = leaveRequestsData.list;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: There is obviously a great deal of stuff going on in this `then()`. Please move it into a documented function, as `loadLeaveRequestsAndCalender` is becoming too long

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 6]
```diff
@@ -354,10 +335,22 @@ define([
...
+
+        _.each(leaveRequests, function (leaveRequest) {
+          leaveRequestsIndexed[leaveRequest.contact_id] = leaveRequestsIndexed[leaveRequest.contact_id] ?
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]:
```js
leaveRequestsIndexed[leaveRequest.contact_id] = leaveRequestsIndexed[leaveRequest.contact_id] || {};
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 7]
```diff
@@ -354,10 +335,22 @@ define([
...
+            leaveRequestsIndexed[leaveRequest.contact_id] : {};
+
+          _.each(leaveRequest.dates, function (leaveRequestDate) {
+            leaveRequestsIndexed[leaveRequest.contact_id][leaveRequestDate.date] = leaveRequest;
+          });
         });
+        return loadCalendar();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: add empty line above

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 8]
```diff
@@ -13,19 +13,21 @@ define([
...
 
 
-  function controller($log, $q, sharedSettings, OptionGroup, Contact, AbsencePeriod, AbsenceType, Calendar, LeaveRequest, PublicHoliday) {
+  function controller($log, $q, $timeout, sharedSettings, OptionGroup, Contact, AbsencePeriod, AbsenceType, Calendar, LeaveRequest, PublicHoliday) {
     $log.debug('Component: manager-leave-calendar');
 
     var dayTypes = [],
       publicHolidays = [],
       leaveRequests = [],
+      leaveRequestsIndexed = [],
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Can you explain what this is for?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 9]
```diff
@@ -488,11 +481,16 @@ define([
     function setCalendarProps(contactID, calendar) {
       var dateObj,
         leaveRequest,
-        dates = Object.keys(calendar.days);
+        dates = Object.keys(calendar.days),
+        indexedByMonth = angular.copy(vm.months.map(function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: You are returning an array of arrays? `[[], [], [], []]`? What's the need for `angular.copy`? Can't it be just
```js
indexedByMonth = vm.months.map(function () { return []; });

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 10]
```diff
@@ -488,11 +481,16 @@ define([
...
+          return [];
+        }));
 
       dates.forEach(function (date) {
         dateObj = calendar.days[date];
-        leaveRequest = getLeaveRequestByDate(contactID, dateObj.date);
+        indexedByMonth[parseInt(moment(dateObj.date).month())].push(dateObj);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: momentjs `month()` already returns a number (https://momentjs.com/docs/#/get-set/month/), is the `parseInt` needed?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 11]
```diff
@@ -488,11 +481,16 @@ define([
...
+        leaveRequest = leaveRequestsIndexed[contactID] ? leaveRequestsIndexed[contactID][dateObj.date]: null;
+        leaveRequest = (leaveRequest && leaveRequest.contact_id === contactID) ? leaveRequest : null;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: This looks pretty bad, tbh. You are doing some convoluted two-steps assignment that is pretty hard to follow. I don't know if it's a matter of how the `leaveRequestsIndexed` is named, or how it is structured (`leaveRequestsIndexed[contactID][dateObj.date]`), but it doesn't come naturally to follow. Please explain it and let's figure out a way to improve this

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 12]
```diff
@@ -488,11 +481,16 @@ define([
...
       var dateObj,
         leaveRequest,
-        dates = Object.keys(calendar.days);
+        dates = Object.keys(calendar.days),
+        indexedByMonth = angular.copy(vm.months.map(function () {
+          return [];
+        }));
 
       dates.forEach(function (date) {
         dateObj = calendar.days[date];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: This could be streamlined and made it simpler

```js
function setCalendarProps(contactID, calendar) {
  var indexedByMonth = vm.months.map(function () { return []; });

  _.each(calendar.days, function (dateObj) {
    var leaveRequest;

    // ...
  });

  return indexedByMonth;
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 13]
```diff
@@ -509,7 +507,31 @@ define([
...
+    /**
+     * Show month loader for all months initially
+     * then hide each loader on the interval of an offset value
+     */
+    function showMonthLoader() {
+      var offset = 0;
+
+      vm.months.forEach(function (month, index) {
+        //show loader
+        vm.loading.months[index] = (index !== vm.months.indexOf(vm.selectedMonths[0]));
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Can you explain this logic? not sure i get it. What is
```js
vm.months.indexOf(vm.selectedMonths[0])
```
supposed to return?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 14]
```diff
@@ -509,7 +507,31 @@ define([
...
+
+        //do not hide the month which is selected by default i.e. the current month
+        if(index !== vm.months.indexOf(vm.selectedMonths[0])) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: `if () {`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 15]
```diff
@@ -509,7 +507,31 @@ define([
...
+          (function (index) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: what's the need for an IIFE?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 16]
```diff
@@ -509,7 +507,31 @@ define([
...
+            $timeout(function () {
+              vm.loading.months[index] = false;
+            }, offset);
+          })(index);
+
+          offset += 500;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: `500` should be in a variable in the controller. Also, have you tried a lower offset?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-calendar_test.js [hunk 17]
```diff
@@ -418,11 +399,39 @@
         return moment(date, sharedSettings.serverDateFormat).clone();
       }
 
-      function getDate(dayType) {
+      function getDate(contact, dateStr) {
+        var date;
+
+        _.each(contact.calendarData, function (month) {
+          _.each(month, function (dateObj) {
+            if(dateObj.date == dateStr) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-calendar_test.js]: `if () {`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 16 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Initially I tried with `2000ms` which seemed fine but that much delay was not necessary, changing it to `500ms` also caused some minor freeze on the loader, where the loader is not spinning when rendering is happening. So `500ms` is optimal.

Moved to variable

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 15 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Checked that its not required, removed it.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 8 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: It is not required anymore, removed.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 3 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Please refer point no 3 of description

> In loadCalendar instead of updating vm.managedContacts on completion of each contact's promise. Calendar data saved temporarily in tempContactData and once all the promises are resolved, data is transferred to vm.managedContacts. This is done so that browser paints the UI only once.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 4 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: It was wrong, moved it to `loadLeaveRequestsAndCalender` function

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 11 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: There was some unnecessary code, now I have
`//fetch leave request, first search by contact_id then by date
        leaveRequest = leaveRequests[contactID] ? leaveRequests[contactID][dateObj.date] : null;`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 13 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Here I am setting loader to true for all months expect for the current month which is selected by default. Ss this function is only called initially, there will a single items in `selectedMonths`, so I am not setting loader to true for current month(`vm.months.indexOf(vm.selectedMonths[0])` is the index of current month)

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 18]
```diff
@@ -304,15 +308,14 @@ define([
...
      */
     function loadCalendar() {
-      var promises = [];
-      _.each(vm.managedContacts, function (contact, index) {
-        promises.push(Calendar.get(contact.id, vm.selectedPeriod.id)
-          .then(function (calendar) {
-            vm.managedContacts[index].calendarData = setCalendarProps(vm.managedContacts[index].id, calendar);
-          }));
-      });
+      tempContactData = angular.copy(vm.managedContacts);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: you already have `_` at your disposal, do you need to inject `angular` also? Isn't `_.clone()` enough?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 19]
```diff
@@ -509,7 +519,33 @@ define([
...
+    function showMonthLoader() {
+      var monthLoadDelay = 500,
+        offset = 0,
+        currentMonthIndex = _.findIndex(vm.months, function (monthObj) {
+          return monthObj.label === vm.selectedMonths[0];
+        });
+
+      vm.months.forEach(function (month, index) {
+        //show loader
+        vm.months[index].loading = (index !== currentMonthIndex);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: `month` is already a reference to `vm.months[index]`
```js
month.loading = (index !== currentMonthIndex)
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 20]
```diff
@@ -509,7 +519,33 @@ define([
...
+
+        //do not hide the month which is selected by default i.e. the current month
+        if (index !== currentMonthIndex) {
+          $timeout(function () {
+            vm.months[index].loading = false;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]:
```js
month.loading = false
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 21]
```diff
@@ -509,7 +519,33 @@ define([
...
 
-      return calendar;
+      return indexedByMonth;
+    }
+
+    /**
+     * Show month loader for all months initially
+     * then hide each loader on the interval of an offset value
+     */
+    function showMonthLoader() {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Simplified
```js
function showMonthLoader() {
  var monthLoadDelay = 500, offset = 0;

  vm.months.forEach(function (month) {
    var isCurrentMonth = month.label === vm.selectedMonths[0];

    // immediately show the current month, delay the others at regular intervals
    if (isCurrentMonth) {
      month.loading = !isCurrentMonth;
    } else {
      $timeout(function () {
        month.loading = false;
      }, offset);

      offset += monthLoadDelay;
    }
  });
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 22]
```diff
@@ -13,18 +14,19 @@ define([
...
+  function controller($log, $q, $timeout, sharedSettings, OptionGroup, Contact, AbsencePeriod, AbsenceType, Calendar, LeaveRequest, PublicHoliday) {
     $log.debug('Component: manager-leave-calendar');
 
     var dayTypes = [],
-      publicHolidays = [],
-      leaveRequests = [],
+      leaveRequests = {},
       leaveRequestStatuses = [],
+      publicHolidays = [],
+      tempContactData = [],
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Please comment this to explain what it is for
